Code/Common/include/sitkCompositeTransform.h

```diff
@@ -0,0 +1,191 @@
+*
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*         http://www.apache.org/licenses/LICENSE-2.0.txt
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+*
+*=========================================================================*/
+
+
+#ifndef sitkCompositeTransform_h
+#define sitkCompositeTransform_h
+
+#include "sitkCommon.h"
+#include "sitkTransform.h"
+
+namespace itk
+{
+
+
+template <typename TParametersValueType, unsigned int NInputDimensions, unsigned int NOutputDimensions>
+class Transform;
+
+namespace simple
+{
+
+/** \class CompositeTransform
+ * \brief This class contains a stack of transforms and concatenates them by composition.
+ *
+ * The transforms are composed in \b reverse order with the back being applied
+ * first:
+ *    \f$ T_0 o T_1 = T_0(T_1(x)) \f$
+ * Transforms are stored in a queue, in the following order:
+ *    \f$ T_0, T_1, ... , T_N-1 \f$
+ *
+ * Transforms are added via AddTransform(). This adds the transforms to the back
+ * of the queue.
+ *
+ * The only parameters of the transform at the back of the queue are exposed and
+ * optimizable for registration.
+ *
+ * Inverse:
+ * The inverse transform is created by retrieving the inverse from each
+ * sub transform and adding them to a composite transform in reverse order.
+ * The m_TransformsToOptimizeFlags is copied in reverse for the inverse.
+ *
+ * \sa itk::CompositeTransform
+ */
+class SITKCommon_EXPORT CompositeTransform
+    : public Transform {
+public:
+  using Self = CompositeTransform;
+  using Superclass = Transform;
+
+  /** \brief Construct an empty CompositeTransform
+   *
+   *  The created CompositeTransform is initialized with zero transforms.
+   *  Additional transforms of dimensions can be added.
+   */
+  explicit CompositeTransform(unsigned int dimensions);
+
+  /** \brief Create CompositeTransform converted or holding the transform argument.
+   *
+   * If the Transform is internally a CompositeTransform, a shallow copy to
+   * the internal transform will be made. Otherwise a new CompositeTransform is
+   * constructed which holds the transform argument.
+   *
+   */
+  CompositeTransform( const Transform & );
+
+  /** \brief A lazy copy constructor
+   *
+   * The new SimpleITK object will reference to the same underlying ITK
+   * CompositeTransform. A deep-copy will be made when the object is modified.
+   */
+  CompositeTransform( const CompositeTransform & );
+
+  /** \brief Create a composite from a vector of Transform
+   *
+   * The CompositeTransform is constructed from deep copies of the Transforms.
+   *
```

Comment:
If one of the transforms in the vector is itself a CompositeTransform, is it deep copied or is the flattened version of it deep copied?

Reply:
deep copies of nested composites should occur so the same layout is preserved.